Feature/mathematical operators (#1451)
neo4j/graphql: docs/modules/ROOT/pages/mathematical-operators.adoc

[[mathematical-operators]]
= Mathematical operators

> Mathematical operators are a handy way to update numerical fields in a single roundtrip to the DB.

Reviewer: I'm not sure if `roundtrip` is the best word here. Also, this refers not to any updates but to updates based on their values. Maybe:

```suggestion
Mathematical operators are a handy way to update numerical fields based on their original values in a single DB transaction.
```

> Specfic operators are available on different numerical types (`Int`, `Float`, xref::type-definitions/types.adoc#type-definitions-types-bigint[`BigInt`]).

Reviewer 2: ```suggestion
Specific operators are available on different numerical types (`Int`, `Float`, xref::type-definitions/types.adoc#type-definitions-types-bigint[`BigInt`]).
```

> We chose to keep operators separated by types as some operators are not available to Natural numbers.

Reviewer: This feels redundant, given the statement above

```suggestion
```

Author: Agree!

> == Supported locations

Reviewer: mm, the name of this section is a bit weird. Personally, I think this part could go directly on the top level introduction, without a section. Thoughts?

```suggestion
```

Author: Agree, I don't know if could be worded different also the followed sentence: "Mathematical operators are supported in Update Mutations within these entities"

Mathematical operators are supported in Update Mutations within these entities:

* `Node`
* `Nested Nodes`
* `Relationship properties`
* `Interfaces`

== Int and BigInt operators
For `Int` and `BigInt` types, the following operators are available:

* `_DECREMENT`
* `_INCREMENT`

== Float operators
For `Float` type, the following operators are available:

* `_ADD`
* `_SUBTRACT`
* `_MULTIPLY`
* `_DIVIDE`

> == The social platform schema

Reviewer: Is this an example?

Maybe we should call this section `example`?

Author: It was intended more like a shared scenario to be used in the followed examples like the "Simple increment operation", maybe we could change it to scenario?
```suggestion
== The social platform scenario
```

Reviewer: I'm happy with this being a section, but it should be inside an "Examples" section as it is confusing whether this is part of the examples or documentation

Author: I see what you mean, I agree, it makes sense to me to move this to the "Examples" section 👍

In this section, we use the following GraphQL schema:

[source, graphql, indent=0]
----
type Video {
  id: ID @id
  views: Int
  ownedBy: User @relationship(type: "OWN_VIDEO", properties: "OwnVideo", direction: IN)
}
type User {
  id: ID @id
  ownVideo: [Video!]! @relationship(type: "OWN_VIDEO", properties: "OwnVideo", direction: OUT)
}
interface OwnVideo @relationshipProperties {
  revenue: Float
}
----

=== Simple increment operation

> Let's say that a user visualizes a video, so we want to increment `viewersCount` for that video by 1.

Reviewer 2: ```suggestion
Let's say that a user views a video, so we want to increment `viewersCount` for that video by 1.
```

Unless you specifically want to use the word `visualizes` for some reason, I think `views` works better here.

[source, graphql, indent=0]
----

> mutation {

Author: Sounds great to me!

  updateVideos(
    where: { id: "VideoID" }
    update: { views_INCREMENT: 1 }
  ) {
    videos {
      id
      views
    }
  }
}
----

=== Nested example

> Now, let's say that the video platform wants to reward the user with 0.01 dollars for the visualization.

Reviewer 2: ```suggestion
Now, let's say that the video platform wants to reward the user with 0.01 dollars for viewing the video.
```

To do that, we have to update the relationship property `revenue`.

[source, graphql, indent=0]
----
mutation Mutation {
  updateUsers(
    where: { id: "UserID" },
    update: { ownVideo: [{ update: { edge: { revenue_ADD: 0.01 } } }] }
  ) {
    users {
      id
      ownVideoConnection {
        edges {
          revenue
        }
      }
    }
  }
}
----

== Optional fields

> Operators remain available for optional fields, this means that if a mathematical operator is been used in a field not defined then a GraphQL error is raised.

Reviewer 2: ```suggestion
Operators remain available for optional fields, this means that if a mathematical operator has been used in a field not defined then a GraphQL error is raised.
```
